server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -116,6 +117,40 @@ public T next() {
         }
     }
 
+    public static <T> Iterator<T> forRange(int lowerBoundInclusive, int upperBoundExclusive, IntFunction<T> fn) {
+        assert lowerBoundInclusive <= upperBoundExclusive : lowerBoundInclusive + " vs " + upperBoundExclusive;
+        if (upperBoundExclusive <= lowerBoundInclusive) {
+            return Collections.emptyIterator();
+        } else {
+            return new IntRangeIterator<>(lowerBoundInclusive, upperBoundExclusive, Objects.requireNonNull(fn));
+        }
+    }
+
+    private static final class IntRangeIterator<T> implements Iterator<T> {
+        private final IntFunction<T> fn;
+        private final int upperBoundExclusive;
+        private int index;
+
+        IntRangeIterator(int lowerBoundInclusive, int upperBoundExclusive, IntFunction<T> fn) {
+            this.fn = fn;
+            this.index = lowerBoundInclusive;
+            this.upperBoundExclusive = upperBoundExclusive;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return index < upperBoundExclusive;
+        }
+
+        @Override
+        public T next() {
+            if (index >= upperBoundExclusive) {
+                throw new NoSuchElementException();
+            }
+            return fn.apply(index++);
```

Comment:
Is it incrementing index after running apply?

Reply:
Yes (this is basically just a copy of `ArrayIterator`)